### PR TITLE
Check pT5's in T5 Duplicate Cleaning Before TC

### PR DESCRIPTION
**Pull request by GNiendorf**
This PR fixes the duplicate rate increase issue that occurs when I try to replace the rz chi-squared cut with the DNN.

Current Timing
Screenshot 2024-11-01 at 4 10 54 PM

This PR Timing
Screenshot 2024-11-01 at 4 14 02 PM

### CODE REVIEW

**Comment by slava77** [comment]
> The PR was built and ran successfully in standalone mode. Here are some of the comparison plots.

this result doesn't quite agree with what you showed me on skype.
After a closer look, I misread the part of code that adds T5s to TCs and thought that there was a bug; so, my other comments on skype about having a fix in a different location do not really apply.
This looks more like an optimization/tuning. As is, this update increases the TC creation time (CPU CI; but logically it adds more compute cost) and doesn't really introduce any visible change in the total

It would be good to see a breakdown compared before and after. 
This is not a very obviously the "right" choice.

The change proposed in [9ec471f](https://github.com/SegmentLinking/cmssw/pull/111/commits/9ec471f4e1e03362de63e357a94a22670419cc3b) adds a T5 assigned to a pT5 to be tested in duplicates against other T5s using 5+ hit overlap .OR. with a dR^2<0.001 (in the T5 dup cleaning right after building 7+ hit overlaps are cleaned and only for pairs that start on the same module). This added T5 by itself will never be added to a TC. So, effectively this is expanding the logic of what's done in `crossCleanT5` where those other T5s are tested against pT5s in the same dR^2<0.001 but without a hit match alternative.

Recall also that @YonsiG and @jchismar have also identified that the first iteration T5 cleaning could go over another module loop (not restricting to pairs that start on the same module).

It would be good to understand why the duplicate rate goes up with the DNN+rz. 
Is the `TightCutFlag` still set with a similar logic? Do you have a variant of DNN+rz on github to check?

**Comment by GNiendorf** [comment]
> this result doesn't quite agree with what you showed me on skype. 

As I mentioned on skype, what I showed you was the before/after with the DNN applied for the rz chi-squared cut. It showed the duplicate rate go down to normal levels with this fix applied. Right now the rz chi-squared cut is contributing to the t5 duplicate removal. With the DNN replacing the rz chi-squared for the tightCutFlag, the t5 TC efficiency goes up and the t5 TC fake rate decreases substantially. The increase in efficiency however drives an increase in the duplicate rate, because it seems like no duplicate cleaning between t5's and pT5's is being applied right now?

[PR_103_Continued.pdf](https://github.com/user-attachments/files/17580413/PR_103_Continued.pdf) - Plots with efficiency and fake rate for T5 TC.

**Comment by slava77** [comment]
> Plots that I showed on skype (again noting that the "Master" here is mislabeled, this is the DNN with rz chi-squared replaced as shown in the presentation above.

please share the breakdown plots (dups, eff, FR in categories).
I'm not sure I understand how the increase pT5 dup rate can go down by an improved T5 cleaning done after pT5s are created

**Comment by slava77** [comment]
> I'm not sure I understand how the increase pT5 dup rate can go down by an improved T5 cleaning done after pT5s are created

I guess it's the T5 dup that goes down much lower, but better to see it to be clear.

Is the TightCutFlag with rzChi2 actually 95% now, or is it just in the documentation?

Seeing the DNN+rz code may also help, in case there are some non-trivialities.

**Comment by GNiendorf** [comment]
@slava77 Both t5 and pt5 duplicate rate go down. Maybe I'm misunderstanding this, but if you have some t5's and a pt5 that were both matched to the same sim track, wouldn't both the pt5 and t5 duplicate rate go up? And if you removed the t5's that were duplicates and left the pt5, wouldn't both go down?

**Comment by slava77** [comment]
> Maybe I'm misunderstanding this, but if you have some t5's and a pt5 that are both matched to the same sim track, wouldn't both the pt5 and t5 duplicate rate go up?

ah, right

File: RecoTracker/LSTCore/src/alpaka/Kernels.h [hunk 1]
```diff
@@ -220,15 +220,15 @@ namespace lst {
...
+            if (quintupletsInGPU.isDup[ix] & 1)
               continue;
 
             for (unsigned int jx1 = 0; jx1 < nQuintuplets_lowmod2; jx1++) {
               unsigned int jx = quintupletModuleIndices_lowmod2 + jx1;
               if (ix == jx)
                 continue;
 
-              if (quintupletsInGPU.partOfPT5[jx] || (quintupletsInGPU.isDup[jx] & 1))
+              if (quintupletsInGPU.isDup[jx] & 1)
```
Comment (slava77) [inline RecoTracker/LSTCore/src/alpaka/Kernels.h:231]: Am I guessing roughly correctly that adding T5s from pT5s in `removeDupQuintupletsInGPUBeforeTC` at least doubles the loop size. (in this case the computation at least quadruples)

IIUC, `partOfPT5` both in ix and jx can be skipped: they can't add to dup cleanup. This should speed things up significantly.

Comment (slava77) [reply RecoTracker/LSTCore/src/alpaka/Kernels.h:231]: in the selection below 
I think that it's worth to check also the case with 4 overlapping hits, can be just `minNHitsForDup_T5 = 4` or if some downsides are visible with the same `minNHitsForDup_T5 = 5` but adding  `or (nMatched== minNHitsForDup_T5-1 and dR2 < 0.004)`

**Comment by GNiendorf** [comment]
Current Timing
Screenshot 2024-11-01 at 4 10 54 PM

This PR Timing
Screenshot 2024-11-01 at 4 14 02 PM

**Comment by GNiendorf** [comment]
@slava77 Results from changing minNHitsForDup_T5 from 5 to 4:

TC_fakerate_etacoarsezoom-3
TC_duplrate_etacoarsezoom
TC_base_0_0_eff_etacoarsezoom

**Comment by slava77** [comment]
> @slava77 Results from changing minNHitsForDup_T5 from 5 to 4:

How does the breakdown look like in this case (is T5 dup still visible)?

is this without a loose dR?
the efficiency drops a bit, but the DR looks significantly lower.

**Comment by GNiendorf** [comment]
> is this without a loose dR? the efficiency drops a bit, but the DR looks significantly lower.

I used the wrong root file as a reference, sorry about that! I updated my comment with the correct plots. This is just setting minNHitsForDup_T5 from 5 to 4, no increase in dR.

Looks like the FR decreases, but the DR is the same now.

**Comment by slava77** [comment]
> Looks like the FR decreases, but the DR is the same now. Looks like an improvement overall, should I add it to the PR?

did the DR breakdown (not posted) change at all or no?

I'm not a big fan of going after FR reduction by looking at overlapping objects (even though it can work quite effectively; here the objects are the same length with similar properties; the overlap-based cleaning is perhaps OK when the fake/real imbalance is large).

**Comment by GNiendorf** [comment]
> did the DR breakdown (not posted) change at all or no?

No, no change at all in DR breakdown.

**Review by slava77 — APPROVED** [review]

**Pull request merged**